// @filmgardi/videojs-subtitle-settings/src/components/SubtitleSettingMenuItem.js
import videojs from 'video.js';
import SubtitleSettingMenu from './SubtitleSettingMenu';

// Concrete classes
const MenuItem = videojs.getComponent('MenuItem');

/**
 * Extend vjs menu item class.
 */

let subtitleSettingMenu = {};

class SubtitleSettingMenuItem extends MenuItem {

  /**
	 * Menu item constructor.
	 *
	 * @param {Player} player - vjs player
	 * @param {Object} item - Item object
	 * @param {SettingButton} item - Item object
	 */
  constructor(player, options, subtitleSettingButton) {
    options.label = player.localize(options.label);
    options.title = player.localize(options.title);
    super(player, options);
    this.options = options;
    this.menuButton = subtitleSettingButton;
    this.initial();
  }

  initial() {
    this.createMenu();
  }

  createItems() {
    // const player = this.player_;
    // new MenuItem(player, { label: 'color', selectable: true })

    return [];
  }

  createMenu() {
    subtitleSettingMenu = new SubtitleSettingMenu(this.player_, {}, this.menuButton);

    subtitleSettingMenu.createMenu();
  }

  /**
	 * Click event for menu item.
	 */
  handleClick(event) {
    subtitleSettingMenu.show();
  }
}

SubtitleSettingMenuItem.hide = () => {
  if (subtitleSettingMenu.hide) {
    subtitleSettingMenu.hide();
  }
};
SubtitleSettingMenuItem.show = () => {
  if (subtitleSettingMenu.show) {
    subtitleSettingMenu.show();
  }
};

export default SubtitleSettingMenuItem;
